app/models/manageiq/providers/hawkular/middleware_manager.rb

```diff
@@ -386,10 +386,13 @@ def remove_middleware_datasource(ems_ref)
     end
 
     def emit_middleware_notification(type, op_name, op_arg, ems_ref)
+      mw_deployment = MiddlewareDeployment.find_by_ems_ref(ems_ref)
+      mw_server = mw_deployment.nil? ? MiddlewareServer.find_by_ems_ref(ems_ref) :
```

Comment:
Wait so this `ems_ref` can be for either a `MiddlewareDeployment` or a `MiddlewareServer`?  That should really be passed in as `klass` or something so you don't have to try finds on multiple classes.